### PR TITLE
Upgrade to samsam 3

### PR DESCRIPTION
**Pull request by mantoni**
#### Purpose (TL;DR) - mandatory

~~__WORK IN PROGRESS! DO NOT MERGE__~~

This PR integrates the changes from https://github.com/sinonjs/samsam/pull/49 with Sinon. ~~Since samsam 3 is not yet release, the build is failing.~~

See intructions below on how to verify locally.

#### Changes

- `lib/sinon/util/core/deep-equal.js` is replaced with `samsam.deepEqual`.
- `lib/match.js` now simply exposes `samsam.createMatcher`.
- Removed obsolete files:
    - `lib/sinon/util/core/every.js`
    - `lib/sinon/util/core/iterable-to-string.js`
    - `lib/sinon/util/core/typeOf.js`
- Removed obsolete dependencies:
    - `lodash.get`
    - `type-detect`
- Parameter ordering of `samsam.deepEqual` has changed from `(expected, actual)` to `(actual, expected)`

 #### How to verify - mandatory
1. Check out this branch
2. `npm install`
3. `npm test`

 #### Checklist for author

- [x] `npm run lint` passes
- [x] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

### CODE REVIEW

**Comment by mroderick** [comment]
I love PRs like these!

**Comment by mantoni** [comment]
Samsam 3 was released and this is passing all tests in all environments. Ready for review.

It was required to update `formatio` and `nise` to avoid having two versions of `samsam` in the "production" dependency tree.

We _could_ also wait for the next major version of `referee` to only have samsam 3 in the development dependency tree. However, all tests work fine with the current version too.

File: lib/sinon/call.js [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 var arrayProto = require("@sinonjs/commons").prototypes.array;
 var sinonMatch = require("./match");
```
Comment (fearphage) [inline lib/sinon/call.js]: This name is incorrect now. It's not `sinonMatch` any longer. It's coming from `samsam`. Also why not be explicit:

```js
var match = require("@sinonjs/samsam").createMatcher;
```

File: lib/sinon/match.js [hunk 2]
```diff
@@ -1,428 +1,5 @@
...
-var deepEqual = require("./util/core/deep-equal").use(match); // eslint-disable-line no-use-before-define
-var every = require("./util/core/every");
-var functionName = require("@sinonjs/commons").functionName;
-var get = require("lodash.get");
-var iterableToString = require("./util/core/iterable-to-string");
-var objectProto = require("@sinonjs/commons").prototypes.object;
-var stringProto = require("@sinonjs/commons").prototypes.string;
-var typeOf = require("./util/core/typeOf");
-var valueToString = require("@sinonjs/commons").valueToString;
+var samsam = require("@sinonjs/samsam");
```
Comment (fearphage) [inline lib/sinon/match.js]: Any reason to keep this file at all? Seems like it's obscuring what's really happening.

File: test/util/core/deep-equal-test.js [hunk 3]
```diff
@@ -330,23 +330,19 @@ describe("util/core/deepEqual", function() {
...
-        var matchDeepEqual = deepEqual.use(match);
-
         var spyA = createSpy();
         var matchA = match(spyA);
 
         var spyB = createSpy();
         var matchB = match(spyB);
 
-        matchDeepEqual(matchA, matchB);
+        deepEqual(matchA, matchB);
```
Comment (fearphage) [inline test/util/core/deep-equal-test.js]: Why are we testing that `samsam` works in `sinon`? Do these tests need to be removed?

This applies to all the changed tests. `Samsam` tests should live in that library and not this one.

**Comment by mantoni** [comment]
@fearphage Thank you for reviewing. You're absolutely right. I kept the tests and `match.js` file initially to show that these changes do not break any tests. Now that the change is complete and all tests pass, I feel confident removing them.

**Pull request merged**